File: companion-docker/app/extractors/blackbaud.py
"""Pre-processor for already-exported Blackbaud CSV. Does NOT touch network or SIS auth — pure data transform.

Rules (v3.38.0): mirrors ``companion-tauri/src-tauri/src/extractors/blackbaud.rs``.
"""
from __future__ import annotations

from . import normalize_header


def preprocess_rows(rows: list[dict[str, str]]) -> list[dict[str, str]]:
    return [_preprocess_row(r) for r in rows]


def _preprocess_row(row: dict[str, str]) -> dict[str, str]:
    is_guardian_row = any(
        normalize_header(k) == "relationshiptype" for k in row.keys()
    )
    out: dict[str, str] = {}
    for raw_k, v in row.items():
        norm = normalize_header(raw_k)
        canonical = _map_header(norm, is_guardian_row)
        if canonical == "external_id":
            out["external_id"] = v
        elif canonical == "guardian_external_id":
            out["guardian_external_id"] = v
        elif canonical == "student_external_id":
            out["student_external_id"] = v
        elif canonical == "first_name":
            out["first_name"] = v
        elif canonical == "last_name":
            out["last_name"] = v
        elif canonical == "middle_name":
            out["middle_name"] = v
        elif canonical == "email":
            out["email"] = v
        elif canonical == "phone":
            out["phone"] = v
        elif canonical == "date_of_birth":
            out["date_of_birth"] = _truncate_iso(v)
        elif canonical == "enrollment_date":
            out["enrollment_date"] = _truncate_iso(v)
        elif canonical == "exit_date":
            out["exit_date"] = _truncate_iso(v)
        elif canonical == "guardian_relationship":
            out["guardian_relationship"] = _normalize_rel(v)
        elif canonical == "grade_level":
            out["grade_level"] = v
        elif canonical == "gender":
            out["gender"] = v.strip()
        else:
            out[norm] = v
    return out


def _map_header(norm: str, is_guardian_row: bool) -> str:
    if norm in ("student_firstname", "first_name", "firstname"):
        return "first_name"
    if norm in ("student_lastname", "last_name", "lastname"):
        return "last_name"
    if norm in ("student_middlename", "middle_name", "middlename"):
        return "middle_name"
    if norm == "student_id":
        return "external_id"
    if norm in ("user_id", "constituent_id", "host_id"):
        return "guardian_external_id" if is_guardian_row else "external_id"
    if norm == "student_external_id":
        return "student_external_id"
    if norm in ("student_email", "user_email", "email"):
        return "email"
    if norm in ("student_phone", "user_phone", "phone"):
        return "phone"
    if norm in ("student_dateofbirth", "dateofbirth", "date_of_birth"):
        return "date_of_birth"
    if norm in ("enrollmentdate", "enrollment_date"):
        return "enrollment_date"
    if norm in ("exitdate", "exit_date"):
        return "exit_date"
    if norm == "relationshiptype":
        return "guardian_relationship"
    if norm in ("student_gradelevel", "grade_level", "gradelevel"):
        return "grade_level"
    if norm in ("student_gender", "gender"):
        return "gender"
    return "_passthrough"


def _truncate_iso(s: str) -> str:
    if s is None:
        return ""
    t = s.strip()
    if not t:
        return ""
    # Cut at first T or space.
    cut_at = None
    for sep in ("T", " "):
        i = t.find(sep)
        if i >= 0 and (cut_at is None or i < cut_at):
            cut_at = i
    candidate = t[:cut_at] if cut_at is not None else t
    parts = candidate.split("-")
    if len(parts) != 3 or len(parts[0]) != 4:
        return t
    if not all(p.isdigit() for p in parts):
        return t
    return candidate


_KNOWN_RELS = {
    "mother",
    "father",
    "guardian",
    "stepmother",
    "stepfather",
    "grandmother",
    "grandfather",
    "aunt",
    "uncle",
}


def _normalize_rel(s: str) -> str:
    t = s.strip()
    low = t.lower()
    if low in _KNOWN_RELS:
        return low
    return t

File: companion-tauri/src-tauri/src/extractors/blackbaud.rs
//! Pre-processor for already-exported Blackbaud CSV. Does NOT touch network or SIS auth — pure data transform.
//!
//! Rules (v3.38.0):
//!   * Dot-nested keys flattened: `"Student.FirstName"` → `first_name`,
//!     `"Student.LastName"` → `last_name`, `"Student.Id"` →
//!     `external_id`.
//!   * Date format `YYYY-MM-DDTHH:MM:SS` truncated to `YYYY-MM-DD`.
//!   * Relationship rows: `RelationshipType` ∈ {Mother, Father,
//!     Guardian, Stepmother, Stepfather} → `guardian_relationship`
//!     (lowercased).
//!   * `User.Id` / `Constituent.Id` → `guardian_external_id` when
//!     `RelationshipType` is present on the row.

use std::collections::HashMap;

use super::normalize_header;

pub fn preprocess_rows(rows: Vec<HashMap<String, String>>) -> Vec<HashMap<String, String>> {
    rows.into_iter().map(preprocess_row).collect()
}

fn preprocess_row(row: HashMap<String, String>) -> HashMap<String, String> {
    // First pass: detect whether this is a guardian-relationship row.
    let is_guardian_row = row
        .keys()
        .any(|k| normalize_header(k) == "relationshiptype");

    let mut out: HashMap<String, String> = HashMap::new();
    for (raw_k, v) in row.into_iter() {
        let norm = normalize_header(&raw_k);
        let mapped = map_blackbaud_header(&norm, is_guardian_row);
        match mapped {
            BBField::ExternalId => {
                out.insert("external_id".to_string(), v);
            }
            BBField::GuardianExternalId => {
                out.insert("guardian_external_id".to_string(), v);
            }
            BBField::StudentExternalId => {
                out.insert("student_external_id".to_string(), v);
            }
            BBField::FirstName => {
                out.insert("first_name".to_string(), v);
            }
            BBField::LastName => {
                out.insert("last_name".to_string(), v);
            }
            BBField::MiddleName => {
                out.insert("middle_name".to_string(), v);
            }
            BBField::Email => {
                out.insert("email".to_string(), v);
            }
            BBField::Phone => {
                out.insert("phone".to_string(), v);
            }
            BBField::DateOfBirth => {
                out.insert("date_of_birth".to_string(), truncate_iso_datetime(&v));
            }
            BBField::EnrollmentDate => {
                out.insert("enrollment_date".to_string(), truncate_iso_datetime(&v));
            }
            BBField::ExitDate => {
                out.insert("exit_date".to_string(), truncate_iso_datetime(&v));
            }
            BBField::GuardianRelationship => {
                out.insert(
                    "guardian_relationship".to_string(),
                    normalize_relationship(&v),
                );
            }
            BBField::GradeLevel => {
                out.insert("grade_level".to_string(), v);
            }
            BBField::Gender => {
                out.insert("gender".to_string(), v.trim().to_string());
            }
            BBField::Passthrough => {
                out.insert(norm, v);
            }
        }
    }
    out
}

enum BBField {
    ExternalId,
    GuardianExternalId,
    StudentExternalId,
    FirstName,
    LastName,
    MiddleName,
    Email,
    Phone,
    DateOfBirth,
    EnrollmentDate,
    ExitDate,
    GuardianRelationship,
    GradeLevel,
    Gender,
    Passthrough,
}

fn map_blackbaud_header(norm: &str, is_guardian_row: bool) -> BBField {
    match norm {
        "student_firstname" | "first_name" | "firstname" => BBField::FirstName,
        "student_lastname" | "last_name" | "lastname" => BBField::LastName,
        "student_middlename" | "middle_name" | "middlename" => BBField::MiddleName,
        "student_id" => BBField::ExternalId,
        "user_id" | "constituent_id" | "host_id" => {
            if is_guardian_row {
                BBField::GuardianExternalId
            } else {
                BBField::ExternalId
            }
        }
        "student_external_id" => BBField::StudentExternalId,
        "student_email" | "user_email" | "email" => BBField::Email,
        "student_phone" | "user_phone" | "phone" => BBField::Phone,
        "student_dateofbirth" | "dateofbirth" | "date_of_birth" => BBField::DateOfBirth,
        "enrollmentdate" | "enrollment_date" => BBField::EnrollmentDate,
        "exitdate" | "exit_date" => BBField::ExitDate,
        "relationshiptype" => BBField::GuardianRelationship,
        "student_gradelevel" | "grade_level" | "gradelevel" => BBField::GradeLevel,
        "student_gender" | "gender" => BBField::Gender,
        _ => BBField::Passthrough,
    }
}

/// Truncate an ISO datetime `YYYY-MM-DDTHH:MM:SS[.fff][Z]` to
/// `YYYY-MM-DD`. If the input is already date-only, pass through. If
/// the input is not a recognized ISO format, pass through verbatim.
fn truncate_iso_datetime(s: &str) -> String {
    let t = s.trim();
    if t.is_empty() {
        return String::new();
    }
    // Accept up to the first 'T' or first space.
    let candidate: &str = match t.find('T').or_else(|| t.find(' ')) {
        Some(i) => &t[..i],
        None => t,
    };
    // Validate looks like `YYYY-MM-DD`.
    let parts: Vec<&str> = candidate.split('-').collect();
    if parts.len() != 3 || parts[0].len() != 4 {
        return t.to_string();
    }
    if !parts.iter().all(|p| p.chars().all(|c| c.is_ascii_digit())) {
        return t.to_string();
    }
    candidate.to_string()
}

fn normalize_relationship(s: &str) -> String {
    let t = s.trim();
    let lower = t.to_ascii_lowercase();
    match lower.as_str() {
        "mother" | "father" | "guardian" | "stepmother" | "stepfather"
        | "grandmother" | "grandfather" | "aunt" | "uncle" => lower,
        _ => t.to_string(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn row(pairs: &[(&str, &str)]) -> HashMap<String, String> {
        pairs
            .iter()
            .map(|(k, v)| ((*k).to_string(), (*v).to_string()))
            .collect()
    }

    #[test]
    fn dot_nested_firstname_flattens() {
        let r = row(&[("Student.FirstName", "Bob")]);
        let out = preprocess_rows(vec![r]);
        assert_eq!(out[0].get("first_name").map(String::as_str), Some("Bob"));
    }

    #[test]
    fn iso_datetime_truncates_to_date() {
        let r = row(&[("Student.DateOfBirth", "2010-03-15T08:00:00Z")]);
        let out = preprocess_rows(vec![r]);
        assert_eq!(
            out[0].get("date_of_birth").map(String::as_str),
            Some("2010-03-15")
        );
    }

    #[test]
    fn relationship_row_user_id_becomes_guardian_id() {
        let r = row(&[
            ("RelationshipType", "Mother"),
            ("User.Id", "G-99"),
        ]);
        let out = preprocess_rows(vec![r]);
        assert_eq!(
            out[0].get("guardian_external_id").map(String::as_str),
            Some("G-99")
        );
        assert_eq!(
            out[0].get("guardian_relationship").map(String::as_str),
            Some("mother")
        );
    }

    #[test]
    fn non_relationship_user_id_stays_external_id() {
        let r = row(&[("User.Id", "U-1")]);
        let out = preprocess_rows(vec![r]);
        assert_eq!(out[0].get("external_id").map(String::as_str), Some("U-1"));
        assert!(out[0].get("guardian_external_id").is_none());
    }

    #[test]
    fn unknown_relationship_passes_through_trimmed() {
        let r = row(&[("RelationshipType", "FosterParent")]);
        let out = preprocess_rows(vec![r]);
        assert_eq!(
            out[0].get("guardian_relationship").map(String::as_str),
            Some("FosterParent")
        );
    }

    #[test]
    fn non_iso_date_preserved() {
        let r = row(&[("Student.DateOfBirth", "invalid")]);
        let out = preprocess_rows(vec![r]);
        assert_eq!(
            out[0].get("date_of_birth").map(String::as_str),
            Some("invalid")
        );
    }

    #[test]
    fn deterministic_output() {
        let r = row(&[
            ("Student.FirstName", "A"),
            ("Student.LastName", "B"),
            ("Student.Id", "1"),
        ]);
        let a = preprocess_rows(vec![r.clone()]);
        let b = preprocess_rows(vec![r]);
        assert_eq!(a, b);
    }
}
